### PR TITLE
Refactor data preprocessing + Add ability to save model

### PR DESCRIPTION
**Pull request by ericagreene**
I'm sorry this is a hodgepodge of changes. I don't know what I don't know with TF. Here's what's changed: 

* I put all the data preprocessing into the `WikiData` class. My idea is that I want all the preprocessing to be separate from the model definition / training / evaluation / saving so that I don't need to duplicate code when I load a saved model and run it on new data. Preprocessing test data also requires access to training `VocabularyProcessor`, so I added the ability to load that. This might be the wrong approach, but it works for now. 

* I added a flag for the number of steps. Now you can run `python model.py --train_data=train.csv --predict_data=test.csv --y_class=toxic --train_steps=1000`

* I figured out how to add more evaluation metric. Now it outputs something like this: 
```
INFO:tensorflow:----------Evaluation on Held-Out Data---------
INFO:tensorflow:Train Size: 76680 Test Size: 19171
INFO:tensorflow:Baseline (class distribution): 0.904473
INFO:tensorflow:Accuracy (sklearn): 0.944708
INFO:tensorflow:accuracy: 0.94470817
INFO:tensorflow:auc: 0.8053536
INFO:tensorflow:false_negatives: 705.0
INFO:tensorflow:false_positives: 355.0
INFO:tensorflow:global_step: 1000
INFO:tensorflow:loss: 0.1694761
INFO:tensorflow:mean_per_class_accuracy: 0.8053536
INFO:tensorflow:true_negatives: 16904.0
INFO:tensorflow:true_positives: 1207.0
```

* I updated to the latest version of TensorFlow. This ended up being necessary to add all of the metrics above. 

* I pulled out the code to write predictions on the validation and test set to disk. I'm going to add another script to load a model and use it to label the test data. 

Among solving other problems, this closes https://github.com/conversationai/wikidetox/issues/70.

### CODE REVIEW

File: conversation_classification/kaggle/model.py [hunk 1]
```diff
@@ -36,38 +34,86 @@
...
+    self.x_train, self.x_train_text = None, None
+    self.x_test, self.x_test_text = None, None
+    self.y_train = None
+    self.y_test = None
+    self.vocab_processor = None
+
+    # If test_mode is True, then put all the data in x_test and y_test
+    if test_mode:
+      train_percent = 0
+      seed = None
```
Comment (nthain) [inline conversation_classification/kaggle/model.py]: Why does seed = None in test mode?

File: conversation_classification/kaggle/model.py [hunk 2]
```diff
@@ -36,38 +34,86 @@
...
+    # Either load a VocabularyProcessor or compute one from the training data
+    if test_mode:
+
+      # If test_mode is True and no vocab_processor_path is specified, then
+      # return an error. We shouldn't train a VocabProcessor at test time.
+      if vocab_processor_path is None:
+        tf.logging.error("Loading data in test_mode with no vocab_processor_path")
+        raise ValueError
+
+      self.vocab_processor = self.load_vocab_processor(vocab_processor_path)
```
Comment (nthain) [inline conversation_classification/kaggle/model.py]: This looks like it's contained in the if test_mode clause. I don't think it should be.

File: conversation_classification/kaggle/model.py [hunk 3]
```diff
@@ -127,15 +180,34 @@ def estimator_spec_for_softmax_classification(logits, labels, mode):
...
+      export_outputs=export_outputs
     )
 
   # EVAL Mode
   eval_metric_ops = {
-    'accuracy': tf.metrics.accuracy(labels=labels, predictions=predicted_classes)
+    'accuracy': tf.metrics.accuracy(
+      labels=labels, predictions=predicted_classes),
+    'auc': tf.metrics.auc(labels=labels, predictions=predicted_classes),
+    'mean_per_class_accuracy': tf.metrics.mean_per_class_accuracy(
```
Comment (nthain) [inline conversation_classification/kaggle/model.py]: Isn't this always the same as accuracy as you're treating this as a binary classification problem?

File: conversation_classification/kaggle/model.py [hunk 4]
```diff
@@ -127,15 +180,34 @@ def estimator_spec_for_softmax_classification(logits, labels, mode):
...
+      labels=labels, predictions=predicted_classes, num_classes=MAX_LABEL),
+    'true_negatives': tf.metrics.true_negatives(
```
Comment (nthain) [inline conversation_classification/kaggle/model.py]: Maybe leave a comment that all of these values are computed at a 0.5 threshold.

File: conversation_classification/kaggle/model.py [hunk 5]
```diff
@@ -171,23 +243,16 @@ def main():
       tf.logging.info('Running in verbose mode')
       tf.logging.set_verbosity(tf.logging.DEBUG)
 
-    # Load and split data
-    tf.logging.debug('Loading data {}'.format(FLAGS.train_data))
-    data = WikiData(FLAGS.train_data)
-
-    x_train_text, x_test_text, y_train, y_test \
-      = data.split(TRAIN_PERCENT, FLAGS.y_class, DATA_SEED)
+    if os.path.isdir(FLAGS.model_dir):
```
Comment (nthain) [inline conversation_classification/kaggle/model.py]: Could you use timestamps to version models or at least have overwriting pre-trained models be a flag? In general, holding on to previous versions of models will prove useful.

File: conversation_classification/kaggle/model.py [hunk 6]
```diff
@@ -209,94 +274,83 @@ def main():
...
       [(p['classes'], p['probs'][1]) for p in predicted_test],
       columns=['y_predicted', 'prob']
     )
-    test_out['comment_text'] = x_train_text
-    test_out['y_true'] = y_test
-
-    # Write out predictions and probabilities for test data
-    tf.logging.info("Writing test predictions to {}".format(TEST_OUT_PATH))
-    test_out.to_csv(TEST_OUT_PATH)
+    test_out['comment_text'] = data.x_train_text
```
Comment (nthain) [inline conversation_classification/kaggle/model.py]: I think this should be data.x_test_text

File: conversation_classification/kaggle/model.py (hunk 1 above)
Comment (ericagreene) [reply conversation_classification/kaggle/model.py]: Good point, I don't think there's a reason.

File: conversation_classification/kaggle/model.py (hunk 2 above)
Comment (ericagreene) [reply conversation_classification/kaggle/model.py]: My reasoning was that in training mode, we'll always want to process the vocabulary given the training data. These are not word embeddings, it's just mapping words to unique numbers. 

This is going to get trickier when I add the ability to use pre-trained word embeddings, but I'd rather cross that bridge when I come to it. Let me know if that doesn't make sense, I might be missing something.

File: conversation_classification/kaggle/model.py (hunk 3 above)
Comment (ericagreene) [reply conversation_classification/kaggle/model.py]: Ha, yes. I was just adding all the metrics I could. I'll take it out as it is totally redundant.

File: conversation_classification/kaggle/model.py (hunk 6 above)
Comment (ericagreene) [reply conversation_classification/kaggle/model.py]: Yes, great catch. I don't actually use that field anywhere so I'll delete those lines. I used make predictions on the test data and write it to file, but I've remove that.

File: conversation_classification/kaggle/model.py (hunk 5 above)
Comment (ericagreene) [reply conversation_classification/kaggle/model.py]: I agree, although I'm explicitly saving the model below in a `saved_model` directory which adds a subdirectory with the timestamp of runtime for each model. So the `model` directory does get overwritten each time, but all the final models are saved separately. Which seems ok?

**Comment by ericagreene** [comment]
@nthain Thanks for the review! I made a few updates based on your suggestions.

File: conversation_classification/kaggle/model.py (hunk 2 above)
Comment (nthain) [reply conversation_classification/kaggle/model.py]: That sounds reasonable. Add a piece to the docstring mentioning that it is only utilized in test mode and must be provided for that mode.

File: conversation_classification/kaggle/model.py (hunk 5 above)
Comment (nthain) [reply conversation_classification/kaggle/model.py]: Ah, ok. That should be fine.
